src/type/definition.js

```diff
@@ -478,12 +478,12 @@ export type GraphQLIsTypeOfFn = (
   info: GraphQLResolveInfo
 ) => boolean
 
-export type GraphQLFieldResolveFn<TSource> = (
+export type GraphQLFieldResolveFn<TSource, TResult> = (
   source: TSource,
   args: {[argName: string]: mixed},
   context: mixed,
   info: GraphQLResolveInfo
-) => mixed
+) => TResult
```

Comment:
Im not sure what the sentence `What issue would Flow not catch that it does catch after this change?` means. My english is poor.

If it is means What will be happened if in the future `TSource -> any` be fixed. The explanation is:
The resovler's hierarchy is something like this:
```
ROOT -> A
  A -> B
    B -> BF
        BF -> KK
    B -> B2
  A -> C
    C -> E
    C -> G
```
Ideally the `TResult` can not be typed ,because the Lower level's input type is Upper level's output. The best solution is only needed to type the input for every resovler, but type the result is not bad.
Because at now ,in inner code `TSource` is cast to `any`.  So this is a alternative solution,to link the type system from ROOT to leaf.( A -> B -> BF -> KK).  
If in the future, a new PR make the `TSource` type rightly pass to its children. The explicit type for `result`  will be no longer needed. and in the user's code the explicit typed `result` will be a redundancy. But the redundancy will not make any problem.(Because B is B).